### ISSUE TITLE
Draft design of NAS and compression visualization

### ISSUE THREAD

**Issue by ultmaster** [microsoft/nni#2890]
## Possible Use Scenario

* Weight sharing NAS developers want to examine their sampling.
* Model compression developers want to see which channels/neurons have been pruned.
* Model/General-NAS developers want to examine their models (PyTorch + tensorboard can do this, but we can do better).
* ~Typical NN visualization (feature maps, gradients, ...)~

## User Experience

1. User writes a model/NAS/compression solution and exports some data (Graph IR+tabular) for visualization.

```python
graph_json = gen_pytorch_graph(model, dummy_input=torch.randn(2, 3, 3, 3))

# supernet with sampling
for i in range(100):
  architecture = mutator.reset()
  writer.add_supernet_path(architecture)

# evolution algorithms: in strategy
for i in range(100):
  mutator = Mutator(base_graph)
  graph = mutator.apply(sampler)
  writer.add_graph_mutation(graph, base_graph)

# compression
for i in range(100):
  writer.add_masks(model)
```

TBD: Stores data each type as a single csv file and put all of them into a folder named `run` (like SummaryWriter).

2. User can choose whatever tools they want to visualize it.
  * Use graphviz to visualize graph IR.
  * Use our interactive graph visualizer, via jupyter, command line, or standalone web server.
  * Use any plotlib (matplotlib/seaborn/...) to visualize tabular data.

```python
graph = Graph.load(graph_json)
# color_events = [Event.load_supernet_path(graph, e, type='color') for e in supernet_events_list]
# tooltip_events = [Event.load_masks(graph, e, type='tooltip') for e in masks_events_list]
# events = [Event.concat(c, t) for c, t in zip(color_events, tooltip_events)]
visualize_graph_with_supernet_sampling(graph, supernet_events_list)
visualize_graph(graph)  # without slider
visualize_graph(graph, events)  # with slider
```

## Proposal

We provide a visualization library that fundamentally consists of three components.

```
src
  vis -> resource (ts) files for visualization
  sdk
    pynni
      vis -> visualization python sdk
```

* TS library that is responsible for rendering graph.
  * Graph library (to cytoscape)
  * Interactive visualizer that supports collapsible graph and dynamic updates (on attributes and css) with slider.
* Python bindings.
  * Graph library (from PyTorch, to networkx/graphviz)
  * Helper function that converts tabular data to renderable attributes on graph.
  * Bindings to launch the TS visualizer.

## Graph IR

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class Custom(nn.Module):
    def __init__(self):
        super(Custom, self).__init__()
        self.op = nn.Conv2d(4, 4, kernel_size=1)

    def forward(self, x):
        return self.op(x)


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(4, 4, kernel_size=1)
        self.custom = Custom()

    def forward(self, x):
        x1 = self.conv1(x)
        x2 = self.custom(x)
        z = torch.cat([x1, x2], 1)
        z = F.relu(z)
        x, y = torch.split(z, 4, 1)
        return x, y


net = Net()
net(torch.randn(2, 4, 8, 8))
```

```json
{
    "inputs": [ { "name": "x"} ],
    "outputs": [ { "name": "out_x" }, { "name": "out_y" } ],
    "nodes": [
        {
            "name": "conv1",
            "type": "Conv2d",
            "params": {
                "in_channels": 4,
                "out_channels": 4,
                "kernel_size": 1
            }
        },
        {
            "name": "custom",
            "type": "Custom",
            "params": {},
            "subgraph": {
                "inputs": [ { "name": "x" } ],
                "outputs": [ { "name": "out" } ],
                "nodes": [
                    {
                        "name": "op",
                        "type": "Conv2d",
                        "params": {
                            "in_channels": 4,
                            "out_channels": 4,
                            "kernel_size": 1
                        }
                    }
                ],
                "edges": [
                    { "head": "x", "tail": "op" },
                    { "head": "op", "tail": "out" }
                ]
            }
        },
        {
            "name": "cat_1",
            "type": "aten::cat",
            "params": { "dim": 1 }
        },
        {
            "name": "relu_1",
            "type": "aten::relu",
            "params": { "dim": 1 }
        },
        {
            "name": "split_1",
            "type": "aten::split",
            "params": { "split_size": 4, "dim": 1 }
        }
    ],
    "edges": [
        { "head": "x", "tail": "conv1" },
        { "head": "x", "tail": "custom" },
        { "head": "conv1", "tail": ["cat_1", 0] },
        { "head": "custom", "tail": ["cat_1", 1] },
        { "head": "cat_1", "tail": "relu_1" },
        { "head": "relu_1", "tail": "split_1" },
        { "head": ["split_1", 0], "tail": "out_x" },
        { "head": ["split_1", 1], "tail": "out_y" }
    ]
}
```
